### PR TITLE
fix(structFromUnstructured) Unstructured converted to struct with ano…

### PR DESCRIPTION
**Pull request by leileiwan**
/kind bug

```release-note
fixs: https://github.com/kubernetes/kubernetes/issues/86285
```
issue: https://github.com/kubernetes/kubernetes/issues/86285

### CODE REVIEW

**Comment by leileiwan** [comment]
CLA  verify

**Comment by leileiwan** [comment]
CLA

**Comment by caesarxuchao** [comment]
/ok-to-test

Can you add unit tests to demonstrate the fix?

**Comment by liggitt** [comment]
thanks for the PR. please include a test case demonstrating the bug and the fix

File: staging/src/k8s.io/apimachinery/pkg/runtime/converter.go [hunk 1]
```diff
@@ -378,9 +378,10 @@ func structFromUnstructured(sv, dv reflect.Value) error {
 	for i := 0; i < dt.NumField(); i++ {
 		fieldInfo := fieldInfoFromField(dt, i)
 		fv := dv.Field(i)
-
-		if len(fieldInfo.name) == 0 {
-			// This field is inlined.
+		// This field is inlined.
+		// if not found fieldInfo.name of Anonymous field in json,the json is inline mode for this Anonymous field
+		_, ok := sv.Interface().(map[string]interface{})[fieldInfo.name]
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/runtime/converter.go]: I would have expected this to trigger off of the `json` tag in the struct, not off the presence/absence of the field name in the JSON data

**Comment by leileiwan** [comment]
> thanks for the PR. please include a test case demonstrating the bug and the fix

OK，I will try it

**Comment by leileiwan** [comment]
* TestCase code:
```
package examples


import(
	"testing"
	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
	"k8s.io/apimachinery/pkg/runtime"
	"encoding/json"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
)

type MyStruct struct{
	metav1.TypeMeta `json:",inline"`
	StructA  `json:",inline"`
	StructB  
}

type StructA struct{
	A string `json:"A,omitempty"`
}

type StructB struct{
	B string `json:"B,omitempty"`
}


func TestUnstructedConvert(t *testing.T){
	myStruct:=NewMyStruct()
	un,err:=ConvertMyStructToUnstructured(&myStruct)
	if err !=nil{
		t.Errorf("convert struct %v to Unstructured error...",myStruct)
	}

	var newStruct MyStruct
	err = runtime.DefaultUnstructuredConverter.FromUnstructured(un.Object, &newStruct)
	if err !=nil{
		t.Errorf("convert Unstructured %v to MyStruct error...",un.Object)
	}

	if newStruct.A !="A"{
		t.Errorf("Convert StructA failed. expect A is string %v but get %v",myStruct.A,newStruct.A)
	}
	if newStruct.B !="B"{
		t.Errorf("Convert StructB failed. expect B is string %v but get %v",myStruct.B,newStruct.B)
	}
}


func NewMyStruct() MyStruct{
	return MyStruct{
		metav1.TypeMeta{
			Kind: "MyStruct",
			APIVersion: "v1",
		},

		StructA{
			A: "A",
		},
		StructB{
			B: "B",
		},
	}
}


func ConvertMyStructToUnstructured(myStruct *MyStruct) (*unstructured.Unstructured, error) {
	var unstructured unstructured.Unstructured
	b, err := json.Marshal(myStruct)
	if err != nil {
		return nil, err
	}
	

	if err := json.Unmarshal(b, &unstructured); err != nil {
		return nil, err
	}
	return &unstructured, nil
}

```

* Test result:
```
--- FAIL: TestUnstructedConvert (0.00s)
    unit_test.go:44: Convert StructB failed. expect B is string B but get
FAIL
FAIL	github.com/kubeflow/pytorch-operator/examples	0.017s
FAIL
```

* StructA will succeed,But StructB not

File: staging/src/k8s.io/apimachinery/pkg/runtime/converter.go (hunk 1 above)
Comment (leileiwan) [reply staging/src/k8s.io/apimachinery/pkg/runtime/converter.go]: Yes. But when there are no json information for StructB, the SturctB will not convert success. Beacuse len(fieldInfo.name) == 0 will not be satisfied (The  fieldInfo.name is string structB).

**Comment by leileiwan** [comment]
* maybe there are something error in my code ,I should use  function ToUnstructured which convert the struct to Unstructured. The test code can pass after it
```
func ConvertMyStructToUnstructured(myStruct *MyStruct) (*unstructured.Unstructured, error) {
	var unstructured unstructured.Unstructured
	var err error
	unstructured.Object, err = runtime.DefaultUnstructuredConverter.ToUnstructured(myStruct)
	if err != nil {
		return nil, err
	}
	return &unstructured, nil
}
```

**Comment by leileiwan** [comment]
*  I have changed the func ConvertMyStructToUnstructured
    * Before run test case, export KUBE_PATCH_CONVERSION_DETECTOR=false. And in the end, the test case will not pass

```
func ConvertMyStructToUnstructured(myStruct *MyStruct) (*unstructured.Unstructured, error) {
	var unstructured unstructured.Unstructured
	var err error
	unstructured.Object, err = runtime.DefaultUnstructuredConverter.ToUnstructured(myStruct)
	if err != nil {
		return nil, err
	}
	return &unstructured, nil
}
```

**Comment by logicalhan** [comment]
> package examples
> 
> 
> import(
> 	"testing"
> 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
> 	"k8s.io/apimachinery/pkg/runtime"
> 	"encoding/json"
> 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
> )
> 
> type MyStruct struct{
> 	metav1.TypeMeta `json:",inline"`
> 	StructA  `json:",inline"`
> 	StructB  
> }
> 
> type StructA struct{
> 	A string `json:"A,omitempty"`
> }
> 
> type StructB struct{
> 	B string `json:"B,omitempty"`
> }
> 
> 
> func TestUnstructedConvert(t *testing.T){
> 	myStruct:=NewMyStruct()
> 	un,err:=ConvertMyStructToUnstructured(&myStruct)
> 	if err !=nil{
> 		t.Errorf("convert struct %v to Unstructured error...",myStruct)
> 	}
> 
> 	var newStruct MyStruct
> 	err = runtime.DefaultUnstructuredConverter.FromUnstructured(un.Object, &newStruct)
> 	if err !=nil{
> 		t.Errorf("convert Unstructured %v to MyStruct error...",un.Object)
> 	}
> 
> 	if newStruct.A !="A"{
> 		t.Errorf("Convert StructA failed. expect A is string %v but get %v",myStruct.A,newStruct.A)
> 	}
> 	if newStruct.B !="B"{
> 		t.Errorf("Convert StructB failed. expect B is string %v but get %v",myStruct.B,newStruct.B)
> 	}
> }

Any reason why this code isn't just added to your commit?

**Comment by leileiwan** [comment]
Sorry, It's my code error,I should just use the  function ToUnstructured and FromUnstructured. I will use ToUnstructured to instead of the function ConvertMyStructToUnstructured in my test case

**Pull request closed without merging**